internal/mage/util/engine.go

```diff
@@ -58,6 +61,17 @@ insecure-entitlements = ["security.insecure"]
 {{ end -}}
 `
 
+// nvidiaSetupHelper provides the required steps to setup nvidia-container-toolkit:
+const nvidiaSetupHelper = `
```

Comment:
What if I want to use in on my computer for my own pipeline? I'll not have access to this file since it's internal to mage.
Is there another way to handle that? Or should I prepare a container the same way you do it here but on my own.

Reply:
@TomChv Good point, thinking this could be moved to the shim, and the file gets initialized there if GPU access is enabled?

Reply:
I think that could work yeah, give it a try

Reply:
Have been exploring this a bit, would it be better to setup the Nvidia runtime at this level? https://github.com/matiasinsaurralde/dagger/blob/gpu-access-2/core/container.go#L1034

So `WithGPU` is called, Nvidia runtime is setup and we still pass the parameters to shim (we'll always need this to signal GPU visibility to the prestart hook).
I don't see an alternative that doesn't involve installing the Nvidia runtime every time. We previously tried mounting Nvidia runtime files from the host into the container -so that no installation step happens- but turns out to be tricky if the container and the host aren't running similar environments.

Let me know if I misunderstood the scenario you described, still thinking about this.

Reply:
> Have been exploring this a bit, would it be better to setup the Nvidia runtime at this level? https://github.com/matiasinsaurralde/dagger/blob/gpu-access-2/core/container.go#L1034

This would be better but you do not know which image is used by the container, since `nvinda runtime` can also be setup on ubuntu, (as far as I understand), this step will mostly fails except if the base image is correct.
That would become tricky become some part of the setup would be up to the user and some other would be on Dagger side, I think it would create a lack of flexibility.

But with [Zenith](https://github.com/dagger/dagger/pull/5443) we might be able to solve this issue thanks to special GPU environment (think about it as extension) that could be load by the user. This would make it much easier!

Reply:
@TomChv I didn't know about Zenith, just reading about it.

CUDA only supports four image types for now (Ubuntu, UBI, RockyLinux and CentOS) and installation steps for the runtime hook are limited too. A set of instructions work for Ubuntu and the other set for CentOS/RHEL: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#

Wandering if we could try to be smart here and perform some small probe to determine if the container image is running one or the other, e.g. a CentOS/RHEL will contain the `dnf` binary but Ubuntu won't, etc.

Reply:
> What if I want to use in on my computer for my own pipeline? I'll not have access to this file since it's internal to mage.

I'm confused - wouldn't you be able to just `export _EXPERIMENTAL_DAGGER_GPU_SUPPORT=1` and then `./hack/dev`?

Reply:
@TomChv have been revisiting this. I think there's some confussion on the environment where the hook runs:
- If we change the engine's base image to Ubuntu, we only need to install Nvidia Container Runtime at that level (engine container).
- Shim should be able to find the path to `nvidia-container-runtime-hook` in the engine's container, that's all.
- **We don't need to install Nvidia tooling in the Dagger-created containers**, we assume an image that's specified by the user already contains Nvidia dependencies. It could be a CUDA image directly -like `nvidia/11.7.1-base-ubuntu20.04` or `nvidia/11.7.1-base-centos7`- or a custom image created by the user that's based on any of these original CUDA images.
- We don't really need to have different behavior for distro flavors as we have control on which image to use for the engine's container.